Store feature

Maartz opened the pull request:
This PR adds namespaces, versions and templates tables
I've tried to find a way to mitigate atom exhaustion via name generation mechanism.
It also adds compression of templates with zlib.
Scheduled clean up of dangling templates, can be tuned, just for the sake of doing it.
Finally I've bootstrap mnesia dets for templates write on disk.

Maartz committed:
add namespaces, versions and templates tables
find a way to mitigate atom exhaustion via name generation
compression of templates with zlib
scheduled clean up of dangling templates
bootstrap mnesia dets for templates write on disk

niamtokik commented on src/gabarit.erl:
```diff
@@ -1,25 +1,95 @@
 %%%===================================================================
-%%% @doc
```
Please adds your name as author.

Maartz replied:
Done

niamtokik commented on src/gabarit.erl:
```diff
@@ -1,25 +1,95 @@
+%% @param Limit Maximum number of templates allowed in the namespace
+%% @returns ok | {error, Reason}
+%% @end
+create_namespace(Namespace, Limit) ->
```
Can you add specifications?

Maartz replied:
Done

niamtokik reviewed (commented):
I will do a second review during the day.

Maartz committed:
add author

Maartz committed:
add author

niamtokik commented on src/gabarit_sup.erl:
```diff
@@ -2,34 +2,36 @@
-    ChildSpecs = [],
-    {ok, {SupFlags, ChildSpecs}}.
+    %% TODO: which option is better for the application
+    StoreOpts = case application:get_env(gabarit, persistence, none) of
```
I usually prefer to have a specific function to deal with options somewhere else than in the `init/1` function.

niamtokik commented on src/gabarit_sup.erl:
```diff
@@ -2,34 +2,36 @@
+                        [{persistence_module, gabarit_mnesia_persistence}]
+                end,
+
+    ChildSpecs = [
```
same thing with children, if they are not dynamic, creating a function somewhere in the code to spawn them if necessary is always cool. Something like that:

```erlang
-spec gabari_store_spec() -> supervisor:child_spec().
gabarit_store_spec() ->
  #{
    ...
  }
```

niamtokik commented on src/gabarit_store.erl:
```diff
@@ -1,19 +1,477 @@
+-define(DEFAULT_PREFIX_STRING, "gabarit$").
+
+%% State record definition
+-record(state,
```
I don't like the term `state` when creating a record in a specific module. Using the name of the module is better when debugging (its always a mess if you have N processes with records called `state`):

```erlang
-record(?MODULE, { ... }).
```

Then to use it:

```erlang
State = #?MODULE{ ... }.
% or
#?MODULE{ key = Value } = State.
```

nitramwilliam replied:
That's indeed quite smart, it'd bring a good hygiene when it comes to debug.
Didn't thought about this kind of case.

niamtokik replied:
Yeah, at first I was using `state` as name, but in production environment when something goes wrong (and it happens a lot), it's easier to have a structure called with the name of the module. This is the same for ets table and so on. Trying to regroup everything under the same banner.

niamtokik commented on src/gabarit_store.erl:
```diff
@@ -1,19 +1,477 @@
+
+  MaxVersionsToKeep = application:get_env(gabarit, max_versions_to_keep, 5),
+
+  lists:foreach(fun({TemplateId, _Namespace, _Path, CurrentVersion}) ->
```
Instead of creating a lambda function, create a full function with a bit of documentation. It will be easier to read.

```erlang
perform_cleanup_foreach({TemplateId, Namespace, Path, CurrentVersion}) ->
  % ...
  ok.

perform_cleanup() ->
  lists:foreach(fun perform_cleanup_foreach/1,
  % ...
```

niamtokik commented on src/gabarit_store.erl:
```diff
@@ -1,19 +1,477 @@
+                       VersionsToDelete =
+                         lists:filter(fun({V, _}) ->
+                                         V =/= CurrentVersion
+                                         andalso length(lists:takewhile(fun({Vx, _}) -> Vx =/= V
```
Create functions to store small lambda functions. Again, easier to read, comment and understand.

niamtokik reviewed (changes requested):
The code seems good, I think I will test it more deeply very soon. ;)

Maartz committed:
switch from state to ?MODULE

state being a way to generic name, it can be very hard to debug while
everything is labeled `state`.
thus giving state in fact the actual module name makes way more sense
and ease the debug process by letting us know the source of the record.

Maartz committed:
reduce function size and leverage composition

Maartz committed:
remove dangling comments and reformat

Maartz committed:
dedicate functions to lean the init

as most of the code is static, delegating its content to dedicated
function is necessary

Maartz committed:
add types and specs

Maartz committed:
identifier seems to be a reserved keyword

Maartz committed:
define types and specs

Maartz commented:
Done

niamtokik commented on src/gabarit.erl:
```diff
@@ -1,25 +1,97 @@
+%% @param Limit Maximum number of templates allowed in the namespace
+%% @returns ok | {error, Reason}
+%% @end
+create_namespace(Namespace, Limit) ->
```
can you create `create_namespace/1` with a default limit? Actually, it could even be better to be a map() containing a list of option (for a future usage).

```erlang
create_namespace(Namespace) ->
  create_namespace(Namespace, #{}).
  
 create_namespace(Namespace, Opts) ->
  Limit = maps:get(limit, Opts, 123),
  ...
```

niamtokik commented on src/gabarit.erl:
```diff
@@ -1,25 +1,97 @@
+%% @param Namespace The namespace to list templates from
+%% @returns {ok, [{TemplateId, Path}]} | {error, Reason}
+%% @end
+list_templates(Namespace) ->
```
If we have a `list_templates/0` function, it could be also nice to create default namespace called `default`. It will then contain all the templates we are using without create a namespace. It will be easier for the users.

Maartz committed:
add some test to assert the new introduce option map works as expected

Maartz committed:
ensure templates directory exists on application startup

Maartz committed:
enhance gabarit_store with improved defaults and robustness

- Add default namespace with infinite template limit
- Implement start_link/0 for easier initialization with defaults
- Add persistence directory configuration
- Enhance error handling with try/catch blocks
- Add reset_namespace/2 function for easier testing
- Add list_templates/0 to get templates across all namespaces
- Fix bug in delete_old_versions tuple pattern
- Refactor initialization for better maintainability
- Add proper state transitions when resetting namespaces
- Improve timeout handling for long-running operations (Maybe not necessary)

These changes make the template store more robust and easier to use
while providing better defaults for normal operation.

Maartz committed:
enhance Gabarit API with improved namespace management

This commit extends the public API with more robust namespace and
template management functions:

- Add create_namespace/1 with sensible defaults (infinity limit)
- Support map-based options for namespace configuration
- Add list_templates/0 to view templates across all namespaces
- Expose reset_namespace/2 for testing and namespace management
- Fix template creation with proper content and path extraction
- Add comprehensive type specifications for API functions
- Use atoms instead of strings for default namespace names
- Improve template compilation with proper ID handling
